feat: add multi-user support (issue #4956)

Somoru opened the pull request:
## Summary

This PR implements comprehensive multi-user support for Trilium Notes, enabling multiple users to collaborate on the same Trilium instance with role-based access control while maintaining full backward compatibility with existing single-user installations.

## Changes

- Add database migration v234 for multi-user schema
- Implement users, roles, user_roles, and note_shares tables
- Add user management service with CRUD operations
- Implement role-based permission system (Admin/Editor/Reader)
- Add RESTful user management API endpoints
- Update login flow to support username + password authentication
- Maintain backward compatibility with legacy password-only login
- Create default admin user from existing credentials during migration
- Add session management for multi-user authentication
- Include TypeScript type definitions for Node.js globals

**Tests:** 948 passed | 17 skipped (965 total)  
**Build:** Successful (server and client)  
**TypeScript:** Zero errors


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#4956: multi-user support](https://oss.issuehunt.io/repos/92111509/issues/4956)
---
</details>
<!-- /Issuehunt content-->

eliandoran commented on apps/server/src/migrations/0234__multi_user_support.ts:
```diff
@@ -0,0 +1,204 @@
+            FOREIGN KEY (ownerId) REFERENCES users(userId) ON DELETE CASCADE,
+            FOREIGN KEY (sharedWithUserId) REFERENCES users(userId) ON DELETE CASCADE
+        )
+    `);
```
If the tables are not used yet (e.g. in order to be able to share notes), there is no reason to create them.

eliandoran commented on apps/server/src/routes/login.ts:
```diff
@@ -114,6 +116,7 @@ function login(req: Request, res: Response) {
 
     const submittedPassword = req.body.password;
     const submittedTotpToken = req.body.totpToken;
+    const submittedUsername = req.body.username; // New field for multi-user mode
```
There seems to be server-side support for user management, but I don't see any way to create a new user or to login, as user.

Without the front-end functionality to back it up, it would be non-functional.

eliandoran commented on apps/server/src/routes/routes.ts:
```diff
@@ -224,6 +225,15 @@ function register(app: express.Application) {
+    route(GET, "/api/users/:userId", [auth.checkApiAuth, csrfMiddleware], usersRoute.getUser, apiResultHandler);
+    route(PST, "/api/users", [auth.checkApiAuth, csrfMiddleware, auth.checkAdmin], usersRoute.createUser, apiResultHandler);
+    route(PUT, "/api/users/:userId", [auth.checkApiAuth, csrfMiddleware], usersRoute.updateUser, apiResultHandler);
+    route(DEL, "/api/users/:userId", [auth.checkApiAuth, csrfMiddleware, auth.checkAdmin], usersRoute.deleteUser, apiResultHandler);
```
These endpoints need to be used somewhere in the client. Preferably as a separate setting where the CRUD operations can take place.

Somoru commented on apps/server/src/migrations/migrations.ts:
```diff
@@ -6,6 +6,11 @@
+    {
+        version: 234,
+        module: async () => import("./0234__multi_user_support.js")
+    },
```
Fixed!  I've updated APP_DB_VERSION from 233 to 234 in apps/server/src/services/app_info.ts. The migration now triggers automatically on server start.

Addressed in commit ccaabcf93.

Somoru commented on apps/server/src/services/user_management.ts:
```diff
@@ -0,0 +1,421 @@
+        utcDateCreated: user.utcDateCreated,
+        utcDateModified: user.utcDateModified
+    };
+}
```
Great question - this is an architectural decision I made deliberately:

Why users are NOT Becca entities:
• Security: User authentication data should NEVER be synced across instances. Each Trilium instance needs its own isolated user database.
• Becca is for content: The Becca entity model is designed for synchronized content (notes, branches, attributes). User credentials are infrastructure, not content.
• Sync risks: Syncing passwords/hashes across instances would create massive security vulnerabilities.

Future sync support:
When we do implement multi-user sync, it will require:
- Per-user sync credentials on each instance
- User-to-user mappings (Alice on Instance A → Alice_Remote on Instance B)
- Content sync separate from authentication
- Each instance maintains its own user_data table

I've documented this extensively in MULTI_USER.md under "Architecture" > "Why not Becca entities?" and listed sync as a future enhancement with specific requirements.

Does this approach make sense to you, or would you like to discuss alternative architectures?

eliandoran replied:
@Somoru , I understand and there's nothing wrong with that.
However, from your statement I also understand that syncing does not work when multi-user is enabled? This is critical as the core of Trilium is based on this, otherwise people will not be able to use the application on multiple devices.

Somoru commented:
## Alternative Implementation: Permission-Aware Sync

I've implemented an alternative approach to multi-user support that addresses the sync blocker mentioned in the comments. Instead of isolating users into separate databases, this implementation uses a **collaborative model with permission-based filtering**.

### Key Difference: Sync Support

The critical issue raised was that sync doesn't work in multi-user mode. This implementation solves that by filtering sync data based on permissions:

```typescript
// In sync.ts - permission filtering during sync
const userId = cls.get('userId');
if (userId) {
    entityChanges = permissions.filterEntityChangesForUser(userId, entityChanges);
}
```

Users only sync notes they have access to, enabling:
- Multiple users working on shared notes
- Private notes that don't sync to other users
- Granular read/write/manage permissions
- Group-based access control

### What's Implemented

**Database Schema:**
- `users` - User accounts with scrypt password hashing
- `groups` - User groups for permission management  
- `note_ownership` - Tracks note owners
- `note_permissions` - Granular access control (read/write/manage)
- `group_members` - Group membership

**API Endpoints (14 total):**
- Permission management: share, revoke, transfer ownership
- Group management: create, delete, add/remove members
- User management: create, list, update roles

**Core Services:**
- Permission checking and filtering
- Group lifecycle management
- User authentication with timing-safe validation

### Architecture

- **Backward compatible** - existing single-user setups work unchanged
- **Permission-aware sync** - the key differentiator from PR #7441
- **Context propagation** - userId flows through request chain via CLS
- **Ownership tracking** - all notes have owners, existing notes assigned to admin

### Documentation

- `MULTI_USER_README.md` - User guide
- `COLLABORATIVE_ARCHITECTURE.md` - Technical deep-dive
- `ADDRESSING_PR_7441.md` - Detailed comparison with PR #7441

Branch: `feat/multi-user-support`

---

**Copy everything below this line for posting:**

---

I've implemented an alternative approach to multi-user support that addresses the sync blocker mentioned in the comments. Instead of isolating users into separate databases, this implementation uses a **collaborative model with permission-based filtering**.

### Key Difference: Sync Support

The critical issue raised was that sync doesn't work in multi-user mode. This implementation solves that by filtering sync data based on permissions:

```typescript
// In sync.ts - permission filtering during sync
const userId = cls.get('userId');
if (userId) {
    entityChanges = permissions.filterEntityChangesForUser(userId, entityChanges);
}
```

Users only sync notes they have access to, enabling:
- Multiple users working on shared notes
- Private notes that don't sync to other users
- Granular read/write/manage permissions
- Group-based access control

### What's Implemented

**Database Schema:**
- `users` - User accounts with scrypt password hashing
- `groups` - User groups for permission management  
- `note_ownership` - Tracks note owners
- `note_permissions` - Granular access control (read/write/manage)
- `group_members` - Group membership

**API Endpoints (14 total):**
- Permission management: share, revoke, transfer ownership
- Group management: create, delete, add/remove members
- User management: create, list, update roles

**Core Services:**
- Permission checking and filtering
- Group lifecycle management
- User authentication with timing-safe validation

### Architecture

- **Backward compatible** - existing single-user setups work unchanged
- **Permission-aware sync** - the key differentiator from PR #7441
- **Context propagation** - userId flows through request chain via CLS
- **Ownership tracking** - all notes have owners, existing notes assigned to admin

### Documentation

- `MULTI_USER_README.md` - User guide
- `COLLABORATIVE_ARCHITECTURE.md` - Technical deep-dive
- `ADDRESSING_PR_7441.md` - Detailed comparison with PR #7441

Branch: `feat/multi-user-support`

deajan commented:
> @Somoru Could you provide more details on how is it going to work? Currently desktop electron uses the same server app that server web instance. Will there be two server apps? With multi-user support to run it on server? And without multi-user support to ship it in desktop electron app?

AFAIK, the server has admin API and keeps track of users/groups. The desktop app only syncs the subset of what is allowed by the server permissions.

@rom1dep Partial sync is a great suggestion. Is there an issue that tracks that?

For instance, partial sync is implemented per user basis in this PR. I think it wouldn't be much more effort to create another token that would allow to specify sync targets, but that's out of scope here.
Perhaps another bounty ?

Somoru commented:
@contributor Good point about the desktop app overhead - this is a valid architectural concern.

**Current Implementation:**
You're right that the desktop app bundles the server code, and currently the multi-user code is integrated into the core server. This does add overhead to:
- Desktop apps (bundled code)
- Single-user server deployments
- Code complexity even when unused

**The Overhead:**
- ~5 new database tables (users, groups, note_ownership, note_permissions, group_members)
- Permission checking services
- Group management logic
- 14 REST API endpoints

**Trade-offs with Current Approach:**
- Simpler codebase (one unified server)
- Backward compatible (detects single-user scenarios automatically)
- Bloat for desktop/single-user deployments
- Code runs even when multi-user is unused

**Possible Solutions:**

1. **Plugin Architecture** (cleanest but most work):
   - Multi-user as optional plugin
   - Desktop/single-user ships without it
   - Server deployments can install plugin

2. **Build-time Flag** (moderate effort):
   - Compile two server variants: `server-single` and `server-multi`
   - Desktop uses `server-single`
   - Deploy multi-user servers with `server-multi`

3. **Runtime Detection with Tree-Shaking** (less ideal):
   - Current approach but with better dead-code elimination
   - Still ships unused code

Would the maintainers prefer I refactor to a plugin architecture? That would address your concern about keeping multi-user overhead external to the core server. I'm happy to pivot if that's the preferred direction.

deajan commented:
@Somoru 
> Admin UI
Currently there's no admin UI - only REST API endpoints. Admin users would need to use curl/Postman or build a simple frontend. I can add a basic admin panel if that's required for the bounty. Let me know if you'd prefer that.

There is definitly a need for a admin user panel, since that feature needs to be usable for everyone having more than one user (can't imagine a family sharing trilium having tu use curl requests).
I'd go for the admin UI, but indeed @eliandoran will have the last word on this.

@Somoru @contributor Multi-user overhead is fairly small (5 almost empty SQL tables, and permission code).
Sync permission checks could be handled optionally, eg if there is more than one user, only then have sync perform permission checks.
This would reduce code splitting between desktop and server app which is IMO very important to keep the codebase clean and bug-free(TM) for future modifications that won't need to handle multiple different codebases, while keeping desktop app fast as if there weren't multiple users.

deajan commented:
> > @Somoru @contributor Multi-user overhead is fairly small (5 almost empty SQL tables, and permission code).
> 
> @deajan I doubt permission code overhead will be small. Especially read permissions, when you need to to check that for massive amount of notes (search, trees etc)

That's where I propose to bypass permission checks when only one user exists, which obviously happens on desktop client.

Somoru commented:
@deajan @contributor @eliandoran 
Before I make the UI changes to implement Admin dashboard and other functionalities, I just want to know, "Does electron app also need multi-user support" if so, I will have to implement a login screen for that aswell and "not bypass" authentication for it. If not, I will just bypass the multi-user feature for electron app.

deajan commented:
> @deajan @contributor @eliandoran Before I make the UI changes to implement Admin dashboard and other functionalities, I just want to know, "Does electron app also need multi-user support" if so, I will have to implement a login screen for that aswell and "not bypass" authentication for it. If not, I will just bypass the multi-user feature for electron app.

Basically, the electron app needs to identify with the server in order to know which user's notes need to be synced, but it does not need full multi user support since only one user will use it at the time.

My best guess is to:
- Server's first user (password only ATM) is admin
- If "classic" login (ie password) is used on electron client, no behavior changes, just sync everything
- Electron client should have possibility to login as user (existing on server), and sync said user's data

In terms of UX, there should be something like "login as server admin" or "login as user" when starting electron client.
Once user is logged in, there could be a checkbox like "remember me" to avoid user logins on every run, for those who would want this workflow

eliandoran commented:
@Somoru , before implementing I would suggest you take a step back and propose a full end-to-end plan, especially given the login and the sync constraints.

deajan commented:
@Somoru I do understand that we asked you alot of stuff in this thread, but this feature isn't exactly the easiest out there, and will have some major implications. Perhaps you would use more of the bounty if Elian is okay with that.
I think Warp can write a up a summary for you, but keep in mind that it creates logical fallacy sometimes, which is why we devs still exist :)

Looking forward hearing from you.

Somoru commented:
# PR #7441 Current State & Implementation Plan

@deajan @eliandoran thank you for your responses, suggestions and the motivation. please check my plan below. If it suits with you, I will start building and hopefully, will deliver soon

The backend is substantially complete with working authentication, permissions, sync filtering, and group management. All users authenticate with username+password (mandatory, no fallbacks). Username is the primary user identifier throughout the system. The application currently lacks the frontend UI layer required to make these features accessible to users.

The simplified authentication model eliminates complexity: every user must have a username, password is always required, and this serves as the foundation for team generation and collaboration features.

---

## What's Already Implemented

### Backend Services (95% Complete)

#### 1. Authentication (`services/auth.ts`)
- Multi-user authentication middleware
- Session management with userId propagation via Continuation-Local-Storage (CLS)
- Support for both single-user password mode and multi-user username+password mode
- Backward compatibility: Single-user instances default to admin user (userId=1)

#### 2. User Management (`services/user_management_collaborative.ts`)
- User creation with secure scrypt hashing (N=16384, r=8, p=1)
- Credential validation with timing-attack protection
- Password changing functionality
- User listing and retrieval
- User roles: admin and user

#### 3. Permissions (`services/permissions.ts`)
- Fine-grained permission levels: read, write, admin
- Permission resolution: owner implicit admin > direct permissions > group permissions
- Functions for checking access, getting permission levels, filtering changes
- Support for both user and group-based permissions

#### 4. Group Management (`services/group_management.ts`)
- Group creation and deletion
- User group membership management
- Group-based permission inheritance
- "All Users" group for organization-wide sharing

#### 5. Sync with Permission Filtering (`routes/api/sync.ts`)
- Pull sync: `getChanged()` filters entity changes by user permissions
- Push sync: `update()` validates write permissions per note
- Permission filtering happens post-sync retrieval, not in database
- Each user receives only notes they own or have access to
- No credentials synced (stays local per instance)

#### 6. Database Schema (Migration v234)
```sql
-- New tables
users                    -- User accounts with authentication
groups                   -- User groups  
group_members           -- Group membership
note_ownership          -- Note owner tracking
note_permissions        -- Granular access control (read/write/admin)
```

### API Endpoints (20+ routes)

**User Management:**
- `POST /api/users` - Create user (admin only)
- `GET /api/users` - List users
- `PUT /api/users/:id` - Update user (admin only)
- `DELETE /api/users/:id` - Delete user (admin only)

**Group Management:**
- `POST /api/groups` - Create group
- `GET /api/groups` - List groups
- `PUT /api/groups/:id` - Update group
- `DELETE /api/groups/:id` - Delete group
- `POST /api/groups/:id/members` - Add user to group
- `DELETE /api/groups/:id/members/:userId` - Remove user from group
- `GET /api/groups/:id/members` - List group members

**Permission Management:**
- `POST /api/notes/:noteId/share` - Share note with user/group
- `GET /api/notes/:noteId/permissions` - List note permissions
- `DELETE /api/notes/:noteId/permissions/:id` - Revoke permission
- `GET /api/notes/accessible` - List notes user can access
- `GET /api/notes/:noteId/my-permission` - Check user's permission on note
- `POST /api/notes/:noteId/transfer-ownership` - Transfer note ownership

### Authentication Flow
1. All logins require username + password (no password-only fallback)
2. Username is the user identifier throughout the system
3. Username maps to userId in database for permission/ownership checks
4. Session stores userId + username for all subsequent requests
5. All protected routes use `req.session.userId` for permission checks
6. First-time setup: creates default admin user (admin/admin) for initial access
7. Admin creates additional users with their own username+password combinations

---

## What's Missing (Frontend Layer)

### 1. Web Admin Dashboard
**Status:** Not implemented (0%)

The backend APIs exist but there is no UI for them. Users cannot:
- Create new users
- Manage groups
- Assign permissions
- View permission settings
- Transfer note ownership

**Implementation Required:**
- React component (`components/admin/UserManagement.tsx`) with:
  - User list table with create/delete buttons
  - Modal for user creation (username, password, email, role)
  - Modal for editing user roles/status
  - Group management interface
  - Permission sharing interface
  - Bulk permission management

**Integration Points:**
- Add admin panel link to main navigation (only for admin users)
- Call existing API endpoints: `/api/users`, `/api/groups`, `/api/notes/:noteId/share`
- Handle 401/403 responses for permission denials

### 2. Login Form Updates
**Status:** Partially implemented

Login page must enforce username+password for all scenarios (no password-only fallback):
- Username field is always required and visible
- Password field is always required
- Client-side validation ensures both fields present
- Server-side auth validates against users table (no legacy password mode)
- Redirect unauthenticated users to login page
- First-time setup: direct to initial admin account creation

**Current State:**
- Backend has username field support in multi-user mode
- Backward compatibility code for password-only login still exists (remove this)
- EJS template needs to always show username field with proper styling
- Session management updated to store both userId and username

**Changes Needed:**
1. Remove password-only authentication fallback from `login.ts`
2. Make username field always required in login form
3. Update error messages to specify username/password mismatch
4. Add first-time setup flow: create initial admin user if no users exist

### 3. Note Sharing UI
**Status:** Not implemented (0%)

Current workflow requires API calls manually. Need UI components for:
- Share button on note header
- Modal to select user/group for sharing
- Permission level selector (read/write/admin)
- List of current note permissions
- Revoke permission button per entry
- Transfer ownership option

**Integration Points:**
- Call `/api/notes/:noteId/share` endpoint
- Call `/api/notes/:noteId/permissions` to list current sharing
- Call `/api/users` and `/api/groups` to populate share recipient lists

### 4. User Indicator / Session UI
**Status:** Missing entirely (0%)

Need to show:
- Currently logged-in user in application header
- User profile dropdown with logout
- Role indicator (admin vs user)
- Option to change password
- Session timeout warnings (if configured)

**Integration Points:**
- Call `/api/app-info` or new `/api/auth/me` endpoint to get current user
- Show in top-right corner of application

### 5. Desktop App Updates
**Status:** Not implemented (0%)

Electron desktop application needs to support username-based login:
- Login dialog asks for username + password (not just password)
- Credential storage per user by username (not global password)
- Store encrypted credentials locally per user profile in config file
- Add user switcher: logout current user, switch to different username
- Remember last logged-in user per device (optional convenience feature)

**Current State:**
- Desktop uses single-user architecture with global password only
- No concept of switching between usernames
- Password stored globally, not per username/user

**Required Changes:**
1. Update login dialog to require username field
2. Store credentials per username in secure config (encrypted)
3. Add logout + user switch menu option in main window
4. Update auto-login logic to use stored username+password per user
5. Allow switching to different user from within app (logout → different username)

---

## Architecture Decisions

### Simplified Authentication: Username as User Identifier

The implementation uses a straightforward username+password authentication model:

**Model:**
- Every user has a unique username and password
- Username is the primary identifier (human-readable, memorable)
- All logins require both username and password (no exceptions, no fallbacks)
- userId in database maps to username for permissions and ownership
- First-time setup creates a default admin user (admin/admin)
- Admin creates additional users by specifying their username and initial password

**Benefits:**
- No complexity switching between password-only and username+password modes
- Consistent authentication across all platforms (web, desktop, mobile)
- Username serves as natural team/group identifier for collaboration
- Team generation builds on top of this foundation: "users with @domain.com usernames can form teams"
- Simpler mental model: "log in as your username"

**Impact on Sync:**
- Each user identified by username throughout sync process
- Permission filtering uses userId (which maps 1:1 to username)
- No ambiguity about which user is syncing
- Desktop app can remember which user was last logged in

### Why Permission-Aware Filtering, Not Database Isolation?

The implementation uses permission-aware sync filtering rather than separate databases per user:

**Pro:**
- True multi-device sync: Alice's desktop and mobile both sync with server
- Shared notes actually work: Bob gets updates Alice makes, and vice versa
- Simple mental model: everyone works on same database, filtered by permissions
- Matches Google Docs/Notion architecture

**Con:**
- Filtering happens in application memory, not database query
- For users with access to many notes, filtering cost increases
- Requires explicit permission checking on every sync pull


## System Architecture Diagram

```
User A (Desktop)              User B (Web)              User C (Mobile)
     |                              |                           |
     +--------- Sync Protocol -------+----------- Sync Protocol --------+
                                     |
                              Server Database
                              - All notes in one DB
                              - Sync filtering by user
                              - Permission validation
                              
User A sees: Notes A owns + Notes shared with A
User B sees: Notes B owns + Notes shared with B  
User C sees: Notes C owns + Notes shared with C (same as Desktop)
```

All three users can edit shared notes; changes sync back and forth automatically.

---

## Next Steps to Production

1. **Enforce username+password login** - remove password-only fallback from auth
2. **Build first-time setup flow** - create default admin user on fresh install
3. **Update login form** - always show username field with proper validation
4. **Build web admin dashboard** - unblocks user creation (username/password/role)
5. **Add sharing UI** - allows users to collaborate (share notes with usernames)
6. **Add session UI** - shows who's logged in (username) and logout button
7. **Update desktop app** - username+password login with per-user credential storage
8. **Team generation** - build on username foundation (e.g., @domain teams)
9. **Testing & validation** - verify all workflows across platforms
10. **Documentation** - user guides showing username-based workflows

deajan commented:
@Somoru Sorry for the late reply, week was complicated.
So far, the architecture seems fine to me.

Please bear in mind that the existing TOTP auth step must still work with the username/password scheme.
Also, in the admin UI, there must be that model upon user deletion which allows to delete all notes or make them own by another user. Generally speaking, notes ownership changes UI should exist anyways.

@eliandoran Could you have another view to make sure I didn't miss anything ?

deajan commented:
@Somoru Any news on this ?
